File: Week Three/Week 3.R
library(dplyr)
library(ggplot2)


### Example 1: Logarithmic Transformation  
df <- read.csv("gapminder.csv")
hist(df$income)
hist(log(df$income))

# Cars dataset
cars = datasets::cars
head(cars$speed)
log(cars$speed) %>% head()
log1p(cars$speed) %>% head()

### Example 2: Square Root Transformation (CO₂ emissions)
df2 <- read.csv("global.csv")
hist(df2$Total)
hist(sqrt(df2$Total))


# Example 3: Reciprocal Transformation (Cars stopping distance)
ggplot(cars, aes(x = seq_along(dist), y = dist)) +
  geom_point(size = 2) + labs(title = "Stopping Distances")

ggplot(cars, aes(x = seq_along(dist), y = 1/dist)) +
  geom_point(size = 2) + labs(title = "Reciprocal Stopping Distances")

# Example 4: Outlier Handling (Titanic – Fare)
# Read and trim fares
df3 <- read.csv("titanic.csv")
Q1 <- quantile(df3$Fare, 0.25, na.rm = TRUE)
Q3 <- quantile(df3$Fare, 0.75, na.rm = TRUE)
IQR <- Q3 - Q1
lower <- Q1 - 1.5 * IQR
upper <- Q3 + 1.5 * IQR
df_trim <- subset(df3, !is.na(Fare) & Fare >= lower & Fare <= upper)

# Counts
orig_n <- sum(!is.na(df3$Fare))
trim_n <- length(df_trim$Fare)
cat(sprintf("Original n = %d, Trimmed n = %d, Removed = %d\n", orig_n, trim_n, orig_n - trim_n))

# Summary statistics
cat("\nOriginal Data:\n")
print(summary(df3$Fare))
cat("\nTrimmed Data:\n")
print(summary(df_trim$Fare))

# Boxplots side-by-side with appropriate scaling
par(mfrow = c(1, 2), mar = c(4, 3, 4, 1))

# Original - auto scale
boxplot(df3$Fare, horizontal = TRUE, main = "Fare — Original", col = "lightblue",
        xlab = "Fare", las = 1)

# Trimmed - auto scale
boxplot(df_trim$Fare, horizontal = TRUE, main = "Fare — After trimming", col = "lightgreen",
        xlab = "Fare", las = 1)

par(mfrow = c(1, 1))


# Boxplots
boxplot(df3$Fare, main = "Original", col = "lightblue", ylab = "Fare")
boxplot(df_trim$Fare, main = "After trimming", col = "lightgreen", ylab = "Fare")

par(mfrow = c(1, 1))

File: Week Three/week 3.py
# -*- coding: utf-8 -*-
"""
Created on Mon Oct 13 10:24:43 2025

@author: Kimani
"""

### Example 1: Logarithmic Transformation  
"""
**Datasets:**  
- **Income** (Gapminder)  
- **Horsepower** (Cars dataset)
"""
#### In Excel
"""
- **Logarithm:**  
    `=LN(cell)` for natural log  
- **Log1p (log(1+x)):**  
    `=LN(1+cell)`
    """
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Gapminder GDP
df = pd.read_csv("gapminder.csv")
income = df["income"]
sns.histplot(income, kde= True)
plt.show()
# log transformation
sns.histplot(np.log(income), kde= True)    
plt.show()

'''
### Example 2: Square Root Transformation (CO₂ Emissions)

#### By Hand
- Calculate \( \sqrt{x} \) for selected countries.
- Compare variance before and after transformation.

#### Excel
- Use: `=SQRT(cell)` for emissions values.

#### Python
'''
df2 = pd.read_csv("global.csv")    
emissions = df2["Total"]
sns.histplot(emissions, kde= True)
plt.show()
# histogram after transformation
sns.histplot(np.sqrt(emissions), kde= True)
plt.show()

### Example 3: Reciprocal Transformation (Cars Stopping Distance) 
'''
from sklearn.datasets import fetch_openml
cars = fetch_openml(name= "cars1", version= 1,
                    as_frame= True).frame
print(cars.head())

cars = cars.to_csv("cars.csv")
'''
cars = pd.read_csv("cars.csv") 

plt.scatter(range(len(cars["MPG"])), cars["MPG"]) 
plt.title("MPG")  
plt.show()

plt.scatter(range(len(cars["MPG"])),
            1/cars["MPG"])
plt.title("Reciprocal MPG")
plt.show()

### Example 4: Outlier Handling (Titanic – Fare)
df3 = pd.read_csv("titanic.csv")
q1, q3 = df3["Fare"].quantile([0.25, 0.75])
iqr = q3 - q1
lower, upper = q1 - 1.5*iqr, q3 + 1.5*iqr
trimmed = df3[(df3["Fare"] >= lower) & 
              (df3["Fare"] <= upper)]
print(f"Original n = {len(df3)},Trimmed 
      n = {len(trimmed)},Removed = {len(df3)
                            - len(trimmed)}")
fig, axes = plt.subplots(1, 2, figsize=(12, 5), 
                         sharey=True)
sns.boxplot(x=df3["Fare"], ax=axes[0],
            color="lightblue")
axes[0].set_title("Fare — Original")
axes[0].set_xlabel("Fare")

sns.boxplot(x=trimmed["Fare"], ax=axes[1],
            color="lightgreen")
axes[1].set_title("Fare — After trimmingoutliers")
axes[1].set_xlabel("Fare")

plt.tight_layout()
plt.show()
    
    
    
    
    
    
    
    
    
    
